### DISCUSSION TITLE
What's is best practise of self-hosting `micropip` wheel file?

### DISCUSSION THREAD

**Discussion by CNSeniorious000** [pyodide/pyodide#4507] [Q&A] [answered]
I am considering releasing a JavaScript package depending on the `pyodide` package, but I need to load `micropip` in it, and I found that the wheels for `micropip` is not included in the `pyodide` package on npm.

- `https://cdn.jsdelivr.net/pyodide/v0.25.0/full/micropip-0.5.0-py3-none-any.whl`
- `https://cdn.jsdelivr.net/pyodide/v0.25.0/full/packaging-23.1-py3-none-any.whl`

So what's the best practise of including these files? I don't want to hard-code any CDN urls like `jsdelivr` or others.

**Comment by CNSeniorious000** [comment]
Can you include the two `whl` files into the npm package?

Or should I download the full build tar file from the GitHub repo and unpack it during build time or in `postinstall`?

I am just curious about what should I do if I neither want to include these two `.whl` files in my git repo nor using a CDN IndexURL.

**Reply by ryanking13** [reply]
Well, if you really don't want to use CDN index urls, you can download those packages from PyPI ([micropip](https://pypi.org/project/micropip/), [packaging](https://pypi.org/project/packaging/)), as they are just pure Python packages.

**Reply by CNSeniorious000** [reply]
Thanks, that's really a solution to get rid of relying on CDN. But I actually don't resist using CDN that much, I just don't like managing micropip dependencies in my JavaScript code.

<details><summary>For example, I still have a few questions about the version of `micropip`:</summary>
<p>
<ol>
<li>
<code>micropip</code> may release a newer version, should I lock the version of <code>micropip</code> to <code>v0.5.0</code>? Or should I fetch the PyPI registry and get the latest version?
</li>
<li>
If newer <code>micropip</code> don't depends on <code>packaging</code> or depends on some more packages? How can I discover this by script? Should I use a dependency manager to resolve the dependencies of <code>micropip</code>?
</li>
</ol>
</p>
</details> 

I noticed that `pyodide` running in nodejs runtime will automatically download missing packages if `indexURL` is not provided, but in browser runtime this will not happen. Is there an API in `pyodide` to invoke this download strategy?

I just don't want to control the version or the dependency file by myself. If `pyodide` provides a default method to invoke the download strategy like in nodejs runtime, I will be happy to use it.

**Reply by ryanking13** [reply, marked as answer]
> micropip may release a newer version, should I lock the version of micropip to v0.5.0? Or should I fetch the PyPI registry and get the latest version?

Our goal is to make micropip compatible with multiple Pyodide versions, but we do not guarantee that this will be possible.
Therefore, we recommend that you stick with one version unless you absolutely need to use the features that a new micropip version requires.

> If newer micropip don't depends on packaging or depends on some more packages? How can I discover this by script? Should I use a dependency manager to resolve the dependencies of micropip?

I don't think this will happen. We've done a lot of work to reduce the runtime dependencies of micropip, and we try to avoid creating additional dependencies where possible.

> I noticed that pyodide running in nodejs runtime will automatically download missing packages if indexURL is not provided, but in browser runtime this will not happen. Is there an API in pyodide to invoke this download strategy?

While Node.js requires a file download process to load packages from the file system, the browser environment loads all packages over the network in real time in the first place. So, in the browser, you can load the micropip simply by running pyodide.loadPackage("micropip").
